### PR TITLE
Preprocessing

### PR DESCRIPTION
**Pull request by sailormarc**

### CODE REVIEW

File: code/preprocessing.py [hunk 1]
```diff
@@ -14,10 +14,25 @@ def preprocess(df: pd.DataFrame) -> pd.DataFrame:
...
+def clean_dataframe_from_eda(df: pd.DataFrame) -> pd.DataFrame:
+    """clean the dataframe based on EDA findings
+    Args:
+        df (pd.DataFrame): the dataframe to clean
+
+    Returns:
+        pd.DataFrame: cleaned dataframe
+    """
+    df = df.drop(columns=['id'], inplace=False)  # remove id column, it is identical to index
+    df = df[df['Region_Code'].apply(lambda x: x.is_integer())] # remove single non-integer region code (see EDA)
```
Comment (sailormarc) [inline code/preprocessing.py:31]: # TODO: skip, just cast to int

File: code/preprocessing.py [hunk 2]
```diff
@@ -14,10 +14,25 @@ def preprocess(df: pd.DataFrame) -> pd.DataFrame:
...
+    df[["Region_Code", "Policy_Sales_Channel"]] = df[["Region_Code", "Policy_Sales_Channel"]].astype(int) 
+    df["Log_Annual_Premium"] = df["Annual_Premium"].apply(lambda x: np.log(x+1))
```
Comment (sailormarc) [inline code/preprocessing.py:33]: Inefficient: use .map()

File: code/preprocessing.py [hunk 3]
```diff
@@ -38,69 +53,31 @@ def vehicle_age_to_int(vehicle_age_col: pd.Series) -> pd.Series:
     return vehicle_age_col.map(map_vehicle_age)
 
 
-def map_categorical_to_int(df: pd.DataFrame) -> pd.DataFrame:
+def map_categorical_to_int(df: pd.DataFrame, encode_salse_channel: bool = False, sales_channel_threshold: float = 98) -> pd.DataFrame:
```
Comment (sailormarc) [inline code/preprocessing.py:56]: Use just one variable

File: code/preprocessing.py [hunk 4]
```diff
@@ -38,69 +53,31 @@ def vehicle_age_to_int(vehicle_age_col: pd.Series) -> pd.Series:
...
-# 1. handle non int value for region code
-# 2. one-hot encode region code and policy sales channel
-# 3. group policy sales channel - parametrize the threshold, default to 98
-# 4. take log of annual premium
-# 5. maybe add a dummy for low premium (below 3k there seems to be an "atom" in the distribution)
-# 6. add dummy for age group >= 35 (from EDA)
-#
+    # group low frequency sales channels into -1 channel
+    if encode_salse_channel:
+        sales_channel_counts = df["Policy_Sales_Channel"].value_counts()
```
Comment (sailormarc) [inline code/preprocessing.py:72]: Add target+ freq encoding if not 1-hot encoding

**Review by sailormarc — COMMENTED** [review]
Keep working on your skills

File: code/preprocessing.py [hunk 5]
```diff
@@ -38,69 +52,35 @@ def vehicle_age_to_int(vehicle_age_col: pd.Series) -> pd.Series:
     return vehicle_age_col.map(map_vehicle_age)
 
 
-def map_categorical_to_int(df: pd.DataFrame) -> pd.DataFrame:
+def map_categorical_to_int(df: pd.DataFrame, sales_channel_threshold: float = None) -> pd.DataFrame:
```
Comment (meiera-CAOS) [inline code/preprocessing.py:55]: why is it float? don't you cast it to int and plan to pass an int threshold ?

Comment (sailormarc) [reply code/preprocessing.py]: the threshold is on the CDF which is real-valued. I do:
 df["Policy_Sales_Channel"][idx] = [ itself if CDF() on it < threshold, else -1 ]
so threshold is to be compared with a real number (a cumulative probability)

Comment (meiera-CAOS) [reply code/preprocessing.py]: oke i just pushed a faulty thing then :) > will make it back to an Optional[float]

**Pull request merged**